### ISSUE TITLE
Readonly/Not-editable cells styling

### ISSUE THREAD

**Issue by faremal** [olifolkerd/tabulator#3076]
Hi Oli,

I've been using Tabulator for the past months now and it is an amazing feature-rich tool. Thanks a lot for your efforts so far and keep up the excellent work!

**Is your feature request related to a problem? Please describe.**
This feature request is not related to a problem per se however, it would be useful in terms of styling. Specifically, when cells are readonly (not editable), it would be useful to apply universal styling to them, say, they become italic with a gray color. I want to stress on the _universal_ part since, as you described here https://stackoverflow.com/questions/54710809/change-style-of-readonly-cells-in-tabulator, it is possible to apply styling on column definitions. However, this does not suffice to my point of view since cells can become dynamically readonly, for example, through an editable callback (i.e., Optional Editing, in http://tabulator.info/docs/4.8/edit) which means that first, different cells can be readonly per row and second, the style per cell can change dynamically (for example, depending on a value of another cell, the cell can become readonly or not)

**Describe the solution you'd like**
Simply put, I propose to add a class (e.g., "tabulator-cell-not-editable" or "tabulator-cell-readonly") on readonly cells that allows cell specific styling for those cells.

**Describe alternatives you've considered**
I am using editable callback (meaning, in the column definition I have an attribute: _editable: function() { True|False }_. My workaround so far is that the aforementioned callback adds a class to the cell which applies the non-editable style. However, this means that the user will have to click on the cell first in order for the style to be applied.

**Additional context**
I hope the above was clear, but if not, let me know to send you some screenshots or other info

Thanks!
Alex

**Comment by olifolkerd** [comment]
Hey @faremal 

Thanks for getting in touch,

That is an interesting idea, Rather than opting for a "not editable class" i would prefer to take the opposite approach, by default no table cells are editable, so no class would be their default state. therefor a class should be applied if a cell is editable, "tabulator-cell-editable", which would fit with the existing "tabulator-cell-editing" class. Tabulator as no concept of "readonly" so it would not really be applicable in this case.

My concern with your scenario is that the `editable` function is not called until someone tries to edit the cell so there is no way or reliably knowing ahead of time of knowing whether the cell is editable or not *(as the callback could check some variable outside of Tabulator)*. Which would make for a very inconsistent behaviour if i implemented this in the core library

So what im actually wondering is if in this scenario, your best approach would just be to use a row formatter that checks through each of it cells and then applies the necessary classes to the cells. you could then call the `update` function on a row component when a cell in that row that might affect things updates. or call a `redraw` on the whole table if you need to affect all rows.

Cheers

Oli )

**Comment by faremal** [comment]
Hi again,

Thanks for the interest and time spent thinking on the suggested feature!

Regarding the "tabulator-cell-editable" and "tabulator-cell-not-editable", the effects of my proposal would be the same given one uses the ":not()" selector, so this is definitely towards the proposed direction.

Regarding your point on the `editable` function, I understand now that you mentioned it. In my case, since the `editable` callback looks for another var or value in the table, I believe it's up to the user (meaning me in this case) to make the appropriate updates if I dynamically change the _editable_ attribute. From your point of view, it would be more than enough to provide a class as a selector (of editable cells as you suggest) which users manipulate

I actually implemented this hack on a table the past two hours. Specifically, I had all rows take a 'tabulator-cell-not-editable' and then called a function to remove or add the class depending on its _editable_ status. The implementation of the functionality to update the classes (Javascript) was not a bother especially considering that I wished to customise it at some points (one non-editable cell was green, another was red). But the implementation of the class 'tabulator-cell-not-editable' on each row definition was not elegant as you can imagine. An additional use-case of the implementation of an _editable_ or _not-editable_ class is that users gain some capabilities with respect to callbacks, e.g., add dynamic table-level cell-edit callbacks like:
`jQuery(".tabulator-cell-editable:has(input)").on("change", function() { do things })`
or
`jQuery(":not(.tabulator-cell-editable):has(input)").on("change", function() { do things })`

I may be thinking of a narrow use-case; my table has many not-editable fields that are updated automatically and, in an event-driven concept, the change of one not-editable cell (that becomes not-editable dynamically) calls for a change in another cell, an ajax call or another action

Greetings,
Alex

**Comment by olifolkerd** [comment]
Hey @faremal 

If you are making those event bindings outside the table then you may run into trouble. Tabulator uses a virtual DOM therefore only rows that are actually visible exist, rows that are outside of the tables viewport are created and destroyed as you scroll the table. therefor they dont exist when the event bindings are made.

For any sort of change bindings you should be using the [dataChanged callback](http://tabulator.info/docs/4.8/release#callbacks) to monitor when data has actually changed.

Cheers

Oli :)

**Comment by faremal** [comment]
Hi Oli,

Right, I am working with very dynamic yet small tables, so the Virtual Dom never created an obstacle. Thanks for the info!

Greetings,
Alex
